histogram/src/atomic.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{BuildError, Config, Error, Histogram, Snapshot};
+use core::sync::atomic::{AtomicU64, Ordering};
+use std::time::SystemTime;
+
+/// A histogram that uses atomic 64bit counters for each bucket.
+///
+/// Unlike the non-atomic variant, it cannot be used directly to report
+/// percentiles. Instead, a snapshot must be taken which captures the state of
+/// the histogram at a point in time.
+pub struct AtomicHistogram {
+    config: Config,
+    start: SystemTime,
+    buckets: Box<[AtomicU64]>,
+}
+
+impl AtomicHistogram {
+    /// Construct a new atomic histogram from the provided parameters. See the
+    /// documentation for [`crate::Config`] to understand their meaning.
+    pub fn new(p: u8, n: u8) -> Result<Self, BuildError> {
+        let config = Config::new(p, n)?;
+
+        Ok(Self::with_config(&config))
+    }
+
+    /// Creates a new atomic histogram using a provided [`crate::Config`].
+    pub fn with_config(config: &Config) -> Self {
+        let mut buckets = Vec::with_capacity(config.total_buckets());
+        buckets.resize_with(config.total_buckets(), || AtomicU64::new(0));
+
+        Self {
+            config: *config,
+            start: SystemTime::now(),
+            buckets: buckets.into(),
+        }
+    }
+
+    /// Increment the bucket that contains the value by one.
+    ///
+    /// This is a convenience method that uses `Instant::now()` as the time
+    /// associated with the observation. If you already have a timestamp, you
+    /// may wish to use `increment_at` instead.
+    pub fn increment(&self, value: u64) -> Result<(), Error> {
+        self.add(value, 1)
+    }
+
+    /// Increment the bucket that contains the value by some count.
+    ///
+    /// This is a convenience method that uses `Instant::now()` as the time
+    /// associated with the observation. If you already have a timestamp, you
+    /// may wish to use the `add_at` instead.
+    pub fn add(&self, value: u64, count: u64) -> Result<(), Error> {
+        let index = self.config.value_to_index(value)?;
+        self.buckets[index].fetch_add(count, Ordering::Relaxed);
+        Ok(())
+    }
+
+    /// Produce a snapshot from this histogram.
+    pub fn snapshot(&self) -> Snapshot {
```

Comment:
We might need another API where the snapshot is pre-allocated and provided by the caller so we can minimize allocation? Fine to add it in a different PR.

Reply:
`snapshot_into()` would be nice. Will do a follow-on PR.